### DISCUSSION TITLE
losing types when passing fixtures to test methods - VS Code

### DISCUSSION THREAD

**Discussion by soezbay** [pytest-dev/pytest#12138] [Q&A]
Hi,

i cannot find a similar problem on google and co. , so i will try my luck here.
Intellisense does not give any suggestions or autofill / autocompletion when passing a fixture with yield or return to test methods.

Hovering over the fixture name in the paramater on a test method reveals that its returning a 'Tuple Generator' but when used in the method then the type is 'any'

**Python 3.12.2 is used**
pytest-8.0.2, pluggy-1.4.0
**pytest plugins im using:**  asyncio-0.23.6, html-4.1.1, metadata-3.1.1, subtests-0.12.1
**Visual Studio Code python plugin:** https://marketplace.visualstudio.com/items?itemName=ms-python.python

```
import iptools as tools  # own correctly implemented class
import pytest

@pytest.fixture
def setup_a():
    p1 = tools.get_a_ip(0) # gets the first ip
    p2 = tools.get_b_ip(1) # gets the second ip
    
    p1... # intellisense autocomplition/suggestions works when pressing crtl + space

    yield p1, p2
    
    # teardown
    tools.del(p1)
    tools.del(p2)

class TestAPI:
    
     def test_ip(self, setup_a): # setup_a has type: def setup_a() -> Generator[tuple[str, str,]]
        # p1, p2 has type 'any'
        p1, p2 = setup_a 
        
        p1. # no working autocompletion/suggestions
        p2. # same problem
```

I could only fix this when explicitly giving p1 and p2 a type, for example:

```
class TestAPI:
     p1: str
     p2: str
     def test_ip(self, setup_a): # setup_a has type: def setup_a() -> Generator[tuple[str, str,]]
        # p1, p2 has type 'any | str'
        p1, p2 = setup_a 
     
        p1. # partially  working autocomplemetion/suggestions
        p2. # same problem
```

Any options to fix that or is there no feature for this?

**Comment by RonnyPfannschmidt** [comment]
its not quite clear what you are asking for, the posted code is so incomplete that we cannot infer enough details

this seems to be a problem with vscode pytest type handling addons - which are handled by vscode

as far as pain python is concerned, your code has no annotations to help at all

**Reply by soezbay** [reply]
i apologize, i forgot the yield and teardown section, and updated the code and added the plugins to the text
I cannot post the original code for data protection reasons of the company, which would not even be a better example

**Reply by RonnyPfannschmidt** [reply]
ok, i believe the vscode tooling has a bug then - the fixture type of yield fixtures should be unwrapped from the generator - if it isnt then they need to argument that

**Reply by soezbay** [reply]
new installtion of VS Code doesn't have an effect, however this issue isn't in pycharm, i probably need to change my IDE preference towards pycharm..

**Comment by The-Compiler** [comment]
Can you show an *actual* example demonstrating the issue you see? Here, your `setup_a` doesn't actually return anything (which would thus cause an exception), and your proposed fix/workaround also can't work, because `p1` and `p2` are not the same thing as `self.p1` and `self.p2`. As already mentioned, your example is too far away from whatever you're actually seeing to be able to say more.

**Reply by soezbay** [reply]
i apologize, i forgot the yield and teardown section, and updated the code and added the plugins to the text
I cannot post the original code for data protection reasons of the company, which would not even be a better example

There is no issue regarding giving errors back or something. The only issue is that p1 and p2 isnt recognized to which imported package it belongs in the test defs
